### PR TITLE
Add first stab at backend auto expiry

### PR DESCRIPTION
**Pull request by MichaelXavier**
This is for issue #7

So it would be nice if we had access to `State` because it would provide
us with `cookieExpires`, but nothing in the machinery of serversession
makes it available to backends for some reason. Also the tests seem to
indicate that the State may not be available at the time of setting up
storange engines (or may be a chicken-and-egg situation). The best
solution seemed to be to take the settings for absolute and idle timeout
that you're going to give to the state and share them.

Just as a reminder for the motivation on this: currently in
serversession's redis backend, you accumulate sessions
indefinitely. Even after they expire, they will stay in redis. The other
backends seem vulnerable to this too but its probably not nearly a big
deal in something like persistent because:
1. Storage is not as much of a premium in peristent's supported backend
   compared to redis. They are primarily storing to disk.
2. Persistent's backends have good querying primitives so it is simple
   to write jobs to delete old sessions.

Redis must fit all data in memory, so storage is at a premium and in
order to find old keys you'd have to issue the frowned-upon KEYS
command (or SCAN which isn't even suppored by hedis yet) and parse every
session.

This change uses redis' built in expiry mechanism and sets the
expiration every time a write operation is made to a session. Sessions
as they would naturally expire on the client side will now automatically
expire and free memory on the backend.

### CODE REVIEW

File: serversession-backend-redis/src/Web/ServerSession/Backend/Redis/Internal.hs [hunk 1]
```diff
@@ -282,14 +287,15 @@ insertSessionImpl session = do
       transaction $ do
         let sk = rSessionKey sid
         r <- batched (R.hmset sk) (printSession session)
-        -- TODO: R.expireat
+        expireSession session sto
+        -- Set the expiration if applicable
```
Comment (meteficha) [inline serversession-backend-redis/src/Web/ServerSession/Backend/Redis/Internal.hs]: Comment should be above.  But I think it's better to just remove the comment, `expireSession` is clear enough.

File: serversession-backend-redis/src/Web/ServerSession/Backend/Redis/Internal.hs [hunk 2]
```diff
@@ -301,6 +307,8 @@ replaceSessionImpl session = do
         let sk = rSessionKey sid
         _ <- R.del [sk]
         r <- batched (R.hmset sk) (printSession session)
+        -- Set the expiration if applicable
```
Comment (meteficha) [inline serversession-backend-redis/src/Web/ServerSession/Backend/Redis/Internal.hs]: Likewise, remove this comment.

File: serversession-backend-redis/src/Web/ServerSession/Backend/Redis/Internal.hs [hunk 3]
```diff
@@ -318,3 +326,17 @@ throwRS
   => StorageException (RedisStorage sess)
   -> R.Redis a
 throwRS = liftIO . E.throwIO
+
+
+expireSession :: Session sess -> RedisStorage sess -> R.RedisTx ()
```
Comment (meteficha) [inline serversession-backend-redis/src/Web/ServerSession/Backend/Redis/Internal.hs:333]: Instead, write a comment here explaining what it does.

**Comment by meteficha** [comment]
Thanks for looking into this issue!

I agree that this PR is not ideal but it's better than the status quo.  Ping me when you've made the changes I've mentioned and are satisfied yourself that this solution works for you.

**Comment by MichaelXavier** [comment]
@meteficha Made the changes you requested to the comments and also set up the default test suite to use timeouts. It doesn't actually matter to the test suites but it at least exercises the code path.

On monday I'm planning on installing this fork at work and giving it a go so I'll let you know how that goes.

**Comment by meteficha** [comment]
Great, do let me know!

**Comment by MichaelXavier** [comment]
@meteficha looks like we're good to go. Here's what I did:
1. Ran a migration that ran `KEYS ssr:*` and set an `EXPIRE` on each key found. This is probably something that you'll want to put in the release notes so end users won't have a set of sessions that live forever.
2. Deployed the change with configured timeouts.
3. Did some login/logout tests and verified that the sessions that were created were having their expirations set.

This patch when released should constitute a breaking version change.

**Comment by aaronc** [comment]
Is this pull request ready to be merged in? Would really like to see this added.

**Comment by MichaelXavier** [comment]
Its ready on my end.

**Comment by iand675** [comment]
@meteficha any chance this can get merged in? I'm seeing this issue too.

**Comment by iand675** [comment]
@snoyberg apologies for being a squeaky wheel, but would you be willing to review this since it's been languishing for a while?

**Comment by snoyberg** [comment]
I've never actually touched this codebase. @meteficha do you want me to merge this in?

**Comment by iand675** [comment]
@snoyberg, @meteficha doesn't appear to have add any public activity GitHub activity for 3 months. Hopefully everything is alright with him, but I'd hope with this being in the yesodweb org, that there's some process for continued maintenance if the primary developer drops out?

**Comment by snoyberg** [comment]
I certainly _can_ step in and merge if necessary. However, I'm not a Hackage maintainer, so I wouldn't be able to upload to Hackage in any event. Given radio silence from Felipe right now, I'm inclined to merge this as (1) it looks OK to my untrained eyes, and (2) I trust you 👍. However, I don't know how much that's going to help given my inability to upload.

**Comment by aaronc** [comment]
I also noticed that @meteficha seems to be silent on another project that got delisted from stackage: https://github.com/prowdsponsor/esqueleto. Is there any other channel that anyone can try to reach out to him? I can try sending an email. Hope he's okay.

I also am waiting on this pull request and will probably end up forking and working off of that.

**Comment by MichaelXavier** [comment]
Ok it has been way too long and I'm finding myself doing the dependency bump ritual on my fork. Let's presume Felipe will not respond. @snoyberg how about this:

1. We merge the redis session expiry PR.
2. We go through the process of taking over maintainership of the project on hackage. Since this project has already gone under the yesod umbrella, it seems fair to me that whatever scheme you have set up for publishing yesod-related packages (a yesod user? maybe your own personal account or some set of core yesod devs?) should be used here.
3. I don't have any specific goals for the future of this project other than keeping it building with the latest snapshots. However, I'd accept commit bit on this project if there isn't anyone on the yesod team with a vested interest in this project, or I'd be happy to submit any future improvements as PRs.

Let me know if this sounds good and if there's anything I can do to help the transition.

**Comment by snoyberg** [comment]
I'm fine with all of this. How about if I give you commit access now to the repo, and you can take actions as you deem fit?

**Comment by MichaelXavier** [comment]
@snoyberg sounds good. One question though: what of the maintainership of the package on Hackage? I'd consider this thread as suitable documentation that we cannot contact Felipe when requesting package takeover, but in requesting a new set of maintainers, is there a list of people that can upload to the yesod project itself that I can add along with my name to the list? I don't want to myself become a bottleneck for the project that doesn't really belong to me.

**Comment by snoyberg** [comment]
The situation on Hackage is not standardized. It would be nice if there were some kind of team concept there, but oh well. I'd say in general @gregwebs and I should be given Hackage access for Yesod related packages, and we can increase that list as needed.

**Comment by hesselink** [comment]
At @MichaelXavier's request I've added him, @snoyberg and @gregwebs as a maintainer to all these packages.

**Comment by snoyberg** [comment]
Thanks Erik!

**Pull request merged**